In [2]:
import nltk

In [3]:
import sys
import sklearn
print("Python:",sys.version)
print("nltk:",nltk.__version__)
print("sklearn",sklearn.__version__)

Python: 3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)]
nltk: 3.4.5
sklearn 0.21.3


In [4]:
nltk.download()

showing info https://raw.githubusercontent.com/nltk/nltk_data/gh-pages/index.xml


True

### Corpus
A collection of written texts, especially the entire works of a particular author or a body of writing on a particular subject.Corpora is the plural of this.Eg:"The darwinian corpus."

### Lexicon
The vocabulary of a person, language, or branch of knowledge.Eg:English dictionary.

### Token
Each entity that is part of whatever was split up based on rules.For examples,each word is a token when a sentence is "tokenised" into words.Each sentence can also be a token if you tokenised the sentences out of a paragraph.

In [5]:
from nltk.tokenize import sent_tokenize,word_tokenize

text='This is a project made in Jupyter Lab. It is an implementation of natural language processing. This is done entirely in python using the python nltk library.'

print(sent_tokenize(text))

['This is a project made in Jupyter Lab.', 'It is an implementation of natural language processing.', 'This is done entirely in python using the python nltk library.']


In [6]:
print(word_tokenize(text))

['This', 'is', 'a', 'project', 'made', 'in', 'Jupyter', 'Lab', '.', 'It', 'is', 'an', 'implementation', 'of', 'natural', 'language', 'processing', '.', 'This', 'is', 'done', 'entirely', 'in', 'python', 'using', 'the', 'python', 'nltk', 'library', '.']


In [7]:
#removing useless data
from nltk.corpus import stopwords
print(set(stopwords.words('english')))

{"didn't", 'further', 'over', 'it', 'few', 'there', 'those', 'can', 'does', 'doesn', "couldn't", 'too', "weren't", 'from', 'how', 'am', 'itself', "wouldn't", 'will', 'but', 'his', 'more', 'off', 'my', 'theirs', 'wouldn', 'these', 'is', 'such', 'not', 'hadn', 'shouldn', 'himself', 'i', 'under', 'if', 'being', 'their', 'aren', 'do', 'up', "shan't", 'with', 'both', 'what', 'isn', 'to', 'against', 'why', 'than', 'same', 'herself', 'should', 'm', 'which', 'for', "doesn't", 'whom', 'are', "needn't", 'have', 'on', "you'd", 'weren', 'the', 'some', 't', "you'll", 'when', 'mustn', 'all', 'as', 'at', 'o', 'in', 'having', "mightn't", 'been', "aren't", 'yourself', 'mightn', 'out', 'she', 'then', 'her', 's', "you've", 'above', 'yours', 'they', "wasn't", 'didn', 'down', 're', 'themselves', 'you', 've', 'couldn', 'shan', 'now', 'below', 'each', 'did', 'needn', 'hasn', 'or', "haven't", 'again', 'ain', "shouldn't", 'ourselves', 'an', 'we', 'were', 'most', 'wasn', 'because', 'by', 'that', 'd', 'until', '

In [8]:
example='Sample text to show stop words being filtered.'
st_wrd=set(stopwords.words('english'))
word_token=word_tokenize(example)
filtered_sentence=[w for w in word_token if w not in st_wrd]
print(word_token)
print(filtered_sentence)

['Sample', 'text', 'to', 'show', 'stop', 'words', 'being', 'filtered', '.']
['Sample', 'text', 'show', 'stop', 'words', 'filtered', '.']


In [9]:
#Stemming words with NLTK
from nltk.stem import PorterStemmer
ps=PorterStemmer()
example_words=['bike','biker','biking','bikes']
for w in example_words:
    print(ps.stem(w))

bike
biker
bike
bike


In [10]:
text="The difference between a woman's bicycle and other bicycles is the location of the top tube. When women began to ride bikes, they wore long skirts. The manufacturers of bikes changed the location of the top tube to make it easier for women to sit down on them while wearing a skirt. Some utility bicycles are similar, for riders who make frequent stops.."
words=word_tokenize(text)
for w in words:
    print(ps.stem(w))

the
differ
between
a
woman
's
bicycl
and
other
bicycl
is
the
locat
of
the
top
tube
.
when
women
began
to
ride
bike
,
they
wore
long
skirt
.
the
manufactur
of
bike
chang
the
locat
of
the
top
tube
to
make
it
easier
for
women
to
sit
down
on
them
while
wear
a
skirt
.
some
util
bicycl
are
similar
,
for
rider
who
make
frequent
stops..


In [11]:
nltk.download()

showing info https://raw.githubusercontent.com/nltk/nltk_data/gh-pages/index.xml


True

In [12]:
from nltk.corpus import udhr
print(udhr.raw('English-Latin1'))

Universal Declaration of Human Rights
Preamble
Whereas recognition of the inherent dignity and of the equal and inalienable rights of all members of the human family is the foundation of freedom, justice and peace in the world, 

Whereas disregard and contempt for human rights have resulted in barbarous acts which have outraged the conscience of mankind, and the advent of a world in which human beings shall enjoy freedom of speech and belief and freedom from fear and want has been proclaimed as the highest aspiration of the common people, 

Whereas it is essential, if man is not to be compelled to have recourse, as a last resort, to rebellion against tyranny and oppression, that human rights should be protected by the rule of law, 

Whereas it is essential to promote the development of friendly relations between nations, 

Whereas the peoples of the United Nations have in the Charter reaffirmed their faith in fundamental human rights, in the dignity and worth of the human person and in

In [13]:
from nltk.corpus import state_union
from nltk.tokenize import PunktSentenceTokenizer

training_tok=state_union.raw('2005-GWBush.txt')
actual_tok=state_union.raw('2006-GWBush.txt')

In [14]:
print(training_tok)

PRESIDENT GEORGE W. BUSH'S ADDRESS BEFORE A JOINT SESSION OF THE CONGRESS ON THE STATE OF THE UNION
 
February 2, 2005


9:10 P.M. EST 

THE PRESIDENT: Mr. Speaker, Vice President Cheney, members of Congress, fellow citizens: 

As a new Congress gathers, all of us in the elected branches of government share a great privilege: We've been placed in office by the votes of the people we serve. And tonight that is a privilege we share with newly-elected leaders of Afghanistan, the Palestinian Territories, Ukraine, and a free and sovereign Iraq. (Applause.) 

Two weeks ago, I stood on the steps of this Capitol and renewed the commitment of our nation to the guiding ideal of liberty for all. This evening I will set forth policies to advance that ideal at home and around the world. 

Tonight, with a healthy, growing economy, with more Americans going back to work, with our nation an active force for good in the world -- the state of our union is confident and strong. (Applause.) 

Our generati

#### With the training text we can train the Punkt Sentence Tokenizer


In [15]:
custom_tok=PunktSentenceTokenizer(training_tok)

#### We Will now Tokenize the sample text


In [16]:
tokenized=custom_tok.tokenize(actual_tok)

In [17]:
print(tokenized)

["PRESIDENT GEORGE W. BUSH'S ADDRESS BEFORE A JOINT SESSION OF THE CONGRESS ON THE STATE OF THE UNION\n \nJanuary 31, 2006\n\nTHE PRESIDENT: Thank you all.", 'Mr. Speaker, Vice President Cheney, members of Congress, members of the Supreme Court and diplomatic corps, distinguished guests, and fellow citizens: Today our nation lost a beloved, graceful, courageous woman who called America to its founding ideals and carried on a noble dream.', 'Tonight we are comforted by the hope of a glad reunion with the husband who was taken so long ago, and we are grateful for the good life of Coretta Scott King.', '(Applause.)', 'President George W. Bush reacts to applause during his State of the Union Address at the Capitol, Tuesday, Jan.', '31, 2006.', "White House photo by Eric DraperEvery time I'm invited to this rostrum, I'm humbled by the privilege, and mindful of the history we've seen together.", 'We have gathered under this Capitol dome in moments of national mourning and national achievemen

#### Part of speech tagging

In [18]:
def process_content():
    try:
        for i in tokenized[:5]:
            words=nltk.word_tokenize(i)
            tag=nltk.pos_tag(words)
            print(tag)
    except Exception as e:
        print(str(e))
process_content()

[('PRESIDENT', 'NNP'), ('GEORGE', 'NNP'), ('W.', 'NNP'), ('BUSH', 'NNP'), ("'S", 'POS'), ('ADDRESS', 'NNP'), ('BEFORE', 'IN'), ('A', 'NNP'), ('JOINT', 'NNP'), ('SESSION', 'NNP'), ('OF', 'IN'), ('THE', 'NNP'), ('CONGRESS', 'NNP'), ('ON', 'NNP'), ('THE', 'NNP'), ('STATE', 'NNP'), ('OF', 'IN'), ('THE', 'NNP'), ('UNION', 'NNP'), ('January', 'NNP'), ('31', 'CD'), (',', ','), ('2006', 'CD'), ('THE', 'NNP'), ('PRESIDENT', 'NNP'), (':', ':'), ('Thank', 'NNP'), ('you', 'PRP'), ('all', 'DT'), ('.', '.')]
[('Mr.', 'NNP'), ('Speaker', 'NNP'), (',', ','), ('Vice', 'NNP'), ('President', 'NNP'), ('Cheney', 'NNP'), (',', ','), ('members', 'NNS'), ('of', 'IN'), ('Congress', 'NNP'), (',', ','), ('members', 'NNS'), ('of', 'IN'), ('the', 'DT'), ('Supreme', 'NNP'), ('Court', 'NNP'), ('and', 'CC'), ('diplomatic', 'JJ'), ('corps', 'NN'), (',', ','), ('distinguished', 'JJ'), ('guests', 'NNS'), (',', ','), ('and', 'CC'), ('fellow', 'JJ'), ('citizens', 'NNS'), (':', ':'), ('Today', 'VB'), ('our', 'PRP$'), ('nat

##### Index for parts of speech

In [19]:
nltk.help.upenn_tagset()

$: dollar
    $ -$ --$ A$ C$ HK$ M$ NZ$ S$ U.S.$ US$
'': closing quotation mark
    ' ''
(: opening parenthesis
    ( [ {
): closing parenthesis
    ) ] }
,: comma
    ,
--: dash
    --
.: sentence terminator
    . ! ?
:: colon or ellipsis
    : ; ...
CC: conjunction, coordinating
    & 'n and both but either et for less minus neither nor or plus so
    therefore times v. versus vs. whether yet
CD: numeral, cardinal
    mid-1890 nine-thirty forty-two one-tenth ten million 0.5 one forty-
    seven 1987 twenty '79 zero two 78-degrees eighty-four IX '60s .025
    fifteen 271,124 dozen quintillion DM2,000 ...
DT: determiner
    all an another any both del each either every half la many much nary
    neither no some such that the them these this those
EX: existential there
    there
FW: foreign word
    gemeinschaft hund ich jeux habeas Haementeria Herr K'ang-si vous
    lutihaw alai je jour objets salutaris fille quibusdam pas trop Monte
    terram fiche oui corporis ...
IN: preposition or

In [20]:
nltk.download()

showing info https://raw.githubusercontent.com/nltk/nltk_data/gh-pages/index.xml


True

### Chunking with NLTK(Grouping)

In [21]:
'''
+=match 1 or more
?=match 0 or 1 repititions
*=match 0 or more repititions
.=Any character except a new line
'''

'\n+=match 1 or more\n?=match 0 or 1 repititions\n*=match 0 or more repititions\n.=Any character except a new line\n'

In [22]:
training_tok=state_union.raw('2005-GWBush.txt')
actual_tok=state_union.raw('2006-GWBush.txt')

#Training the Punkt tokenizer
custom_tok=PunktSentenceTokenizer(training_tok)

#Tokenizing the text
tokenized=custom_tok.tokenize(actual_tok)

def process_content1():
    try:
        for i in tokenized[:2]:
            words=nltk.word_tokenize(i)
            tagged=nltk.pos_tag(words)
            
            #Combining parts of speech with regular expression
            chunkgram=r"""Chunk:{<RB.?>*<VB.?>*<NNP>*<NN>?}"""
            chunkParser=nltk.RegexpParser(chunkgram)
            chunked=chunkParser.parse(tagged)
            
            #draw the chunks with nltk
            chunked.draw()
    except Exception as e:
        print(str(e))
process_content1()

In [23]:
'''
<RB.?>*="0 or more of any tense adverb"
<VB.?>*="0 or more of any tense verb"
<NNP>*="one or more Proper Noun" followed by
<NN>?="zero or one singular noun"
'''

'\n<RB.?>*="0 or more of any tense adverb"\n<VB.?>*="0 or more of any tense verb"\n<NNP>*="one or more Proper Noun" followed by\n<NN>?="zero or one singular noun"\n'

In [24]:
#Chunking with NLTK

training_tok=state_union.raw('2005-GWBush.txt')
actual_tok=state_union.raw('2006-GWBush.txt')

#Training the Punkt tokenizer
custom_tok=PunktSentenceTokenizer(training_tok)

#Tokenizing the text
tokenized=custom_tok.tokenize(actual_tok)

#Function to tokenize each word with a part of speech.
def process_content2():
    try:
        for i in tokenized[:10]:
            words=nltk.word_tokenize(i)
            tagged=nltk.pos_tag(words)
            
            #Combining parts of speech with regular expression
            chunkgram=r"""Chunk:{<RB.?>*<VB.?>*<NNP>*<NN>?}"""
            chunkParser=nltk.RegexpParser(chunkgram)
            chunked=chunkParser.parse(tagged)
            
            #Printing the nltk tree
            for subtree in chunked.subtrees(filter=lambda t:t.label() == 'Chunk'):
                print(subtree)
    except Exception as e:
        print(str(e))
process_content2()

(Chunk PRESIDENT/NNP GEORGE/NNP W./NNP BUSH/NNP)
(Chunk ADDRESS/NNP)
(Chunk A/NNP JOINT/NNP SESSION/NNP)
(Chunk THE/NNP CONGRESS/NNP ON/NNP THE/NNP STATE/NNP)
(Chunk THE/NNP UNION/NNP January/NNP)
(Chunk THE/NNP PRESIDENT/NNP)
(Chunk Thank/NNP)
(Chunk Mr./NNP Speaker/NNP)
(Chunk Vice/NNP President/NNP Cheney/NNP)
(Chunk Congress/NNP)
(Chunk Supreme/NNP Court/NNP)
(Chunk corps/NN)
(Chunk Today/VB)
(Chunk nation/NN)
(Chunk lost/VBD)
(Chunk beloved/VBN)
(Chunk woman/NN)
(Chunk called/VBD America/NNP)
(Chunk founding/NN)
(Chunk carried/VBD)
(Chunk dream/NN)
(Chunk Tonight/NN)
(Chunk are/VBP comforted/VBN)
(Chunk hope/NN)
(Chunk reunion/NN)
(Chunk husband/NN)
(Chunk was/VBD taken/VBN)
(Chunk so/RB long/RB ago/RB)
(Chunk are/VBP)
(Chunk life/NN)
(Chunk Coretta/NNP Scott/NNP King/NNP)
(Chunk Applause/NNP)
(Chunk President/NNP George/NNP W./NNP Bush/NNP)
(Chunk reacts/VBZ)
(Chunk applause/VB)
(Chunk State/NNP)
(Chunk Union/NNP Address/NNP)
(Chunk Capitol/NNP)
(Chunk Tuesday/NNP)
(Chunk Jan/NNP

In [25]:
#Chinking with NLTK

training_tok=state_union.raw('2005-GWBush.txt')
actual_tok=state_union.raw('2006-GWBush.txt')

#Training the Punkt tokenizer
custom_tok=PunktSentenceTokenizer(training_tok)

#Tokenizing the text
tokenized=custom_tok.tokenize(actual_tok)

#Function to tokenize each word with a part of speech.
def process_content2():
    try:
        for i in tokenized[:10]:
            words=nltk.word_tokenize(i)
            tagged=nltk.pos_tag(words)
            
            #}{ means that multiple verbs,prepositions or the word'TO' is being removed from the chink
            
            #Combining parts of speech with regular expression
            
            chunkgram=r"""Chunk: {<.*>+}            
            }<VB.?|IN|DT|TO>+{"""                   #Removing the verbs,prepositions,determiners or word 'TO'
            chunkParser=nltk.RegexpParser(chunkgram)
            chunked=chunkParser.parse(tagged)
            
            #Printing the nltk tree
            
            for subtree in chunked.subtrees(filter=lambda t:t.label() == 'Chunk'):
                print(subtree)
    except Exception as e:
        print(str(e))
process_content2()

(Chunk PRESIDENT/NNP GEORGE/NNP W./NNP BUSH/NNP 'S/POS ADDRESS/NNP)
(Chunk A/NNP JOINT/NNP SESSION/NNP)
(Chunk THE/NNP CONGRESS/NNP ON/NNP THE/NNP STATE/NNP)
(Chunk
  THE/NNP
  UNION/NNP
  January/NNP
  31/CD
  ,/,
  2006/CD
  THE/NNP
  PRESIDENT/NNP
  :/:
  Thank/NNP
  you/PRP)
(Chunk ./.)
(Chunk
  Mr./NNP
  Speaker/NNP
  ,/,
  Vice/NNP
  President/NNP
  Cheney/NNP
  ,/,
  members/NNS)
(Chunk Congress/NNP ,/, members/NNS)
(Chunk
  Supreme/NNP
  Court/NNP
  and/CC
  diplomatic/JJ
  corps/NN
  ,/,
  distinguished/JJ
  guests/NNS
  ,/,
  and/CC
  fellow/JJ
  citizens/NNS
  :/:)
(Chunk our/PRP$ nation/NN)
(Chunk ,/, graceful/JJ ,/, courageous/JJ woman/NN who/WP)
(Chunk America/NNP)
(Chunk its/PRP$ founding/NN ideals/NNS and/CC)
(Chunk noble/JJ dream/NN ./.)
(Chunk Tonight/NN we/PRP)
(Chunk hope/NN)
(Chunk glad/JJ reunion/NN)
(Chunk husband/NN who/WP)
(Chunk so/RB long/RB ago/RB ,/, and/CC we/PRP)
(Chunk grateful/JJ)
(Chunk good/JJ life/NN)
(Chunk Coretta/NNP Scott/NNP King/NNP ./.)
(Chunk

#### Named Entity recognition usnig nltk's inbuilt named entity recognizer

In [26]:
def process_content3():
    try:
        for i in tokenized[:5]:
            words=nltk.word_tokenize(i)
            tagged=nltk.pos_tag(words)
            named_Ent=nltk.ne_chunk(tagged,binary=False)
            
            #draw the named entity
            named_Ent.draw()
            
    except Exception as e:
        print(str(e))
process_content3()

## Text Classification

#### Text Classification using NLTK

We are now going to classify text using simple machine learning classification algorithms

In [27]:
import random
import nltk
from nltk.corpus import movie_reviews

In [28]:
#building list of reviews
documents=[(list(movie_reviews.words(fileid)),category)
        for category in movie_reviews.categories()
        for fileid in movie_reviews.fileids(category)]
#shuffle the reviews

random.shuffle(documents)
print('Number of reviews:',len(documents))
print('First review:',documents[0])

all_words=[]
for w in movie_reviews.words():
    all_words.append(w.lower())
all_words=nltk.FreqDist(all_words)

print('Most common words:{}',all_words.most_common(15))
print('The word good:',all_words["good"])
print("The word bad:",all_words["bad"])

Number of reviews: 2000
First review: (['the', 'word', "'", 'rest', "'", 'in', 'the', 'title', 'should', 'be', 'stressed', 'since', 'this', 'is', 'a', 'movie', 'that', 'lacks', 'any', 'action', 'scenes', 'at', 'all', '.', 'what', 'does', 'takes', 'place', 'is', 'more', 'based', 'on', 'character', 'development', '.', 'the', 'film', 'starts', 'right', 'in', 'the', 'middle', 'of', 'the', 'lives', 'of', 'four', 'people', ':', 'nick', ',', 'beth', ',', 'carol', 'and', 'sid', '.', 'what', 'they', 'have', 'done', 'earlier', 'is', 'for', 'the', 'most', 'part', 'only', 'hinted', 'at', '.', 'nick', 'and', 'beth', 'have', 'a', 'relationship', 'and', 'have', 'decided', 'to', 'move', 'from', 'enfield', ',', 'arizona', 'to', 'butte', ',', 'montana', '.', '(', 'and', 'i', 'who', 'thought', 'that', 'butte', 'was', 'pronounced', 'slightly', 'different', '.', '.', '.', ')', 'carol', 'is', 'nick', "'", 's', 'ex', '-', 'girlfriend', 'and', 'beth', "'", 's', 'best', 'friend', ',', 'sid', 'is', 'a', 'joint'

In [29]:
len(all_words)

39768

In [30]:
#The words list is scaled down to 5000 on which on ML algorithm will be trained
words_features=list(all_words.keys())[:5000]

In [31]:
# A function is defined which will find which of the 5000 word features are contained in a review
def find_features(document):
    words=set(document)
    features={}
    
    for w in words_features:
        features[w]=(w in words)
    return features

#Example : a negative review
features=find_features(movie_reviews.words('neg/cv000_29416.txt'))
for key,value in features.items():
    if value == True:
        print(key)

plot
:
two
teen
couples
go
to
a
church
party
,
drink
and
then
drive
.
they
get
into
an
accident
one
of
the
guys
dies
but
his
girlfriend
continues
see
him
in
her
life
has
nightmares
what
'
s
deal
?
watch
movie
"
sorta
find
out
critique
mind
-
fuck
for
generation
that
touches
on
very
cool
idea
presents
it
bad
package
which
is
makes
this
review
even
harder
write
since
i
generally
applaud
films
attempt
break
mold
mess
with
your
head
such
(
lost
highway
&
memento
)
there
are
good
ways
making
all
types
these
folks
just
didn
t
snag
correctly
seem
have
taken
pretty
neat
concept
executed
terribly
so
problems
well
its
main
problem
simply
too
jumbled
starts
off
normal
downshifts
fantasy
world
you
as
audience
member
no
going
dreams
characters
coming
back
from
dead
others
who
look
like
strange
apparitions
disappearances
looooot
chase
scenes
tons
weird
things
happen
most
not
explained
now
personally
don
trying
unravel
film
every
when
does
give
me
same
clue
over
again
kind
fed
up
after
while
biggest


In [32]:
#For the rest of the reviews
featuresets=[(find_features(rev),category) for (rev,category) in documents]

In [33]:
from sklearn import model_selection

#define seed for reproducibility
seed=1

#splitting the data into training and testing examples
training,testing= model_selection.train_test_split(featuresets,test_size=0.25, random_state=seed)

In [34]:
print(len(training))
print(len(testing))

1500
500


In [35]:
#sklearn algorithms in nltk
from nltk.classify.scikitlearn import SklearnClassifier
from sklearn.svm import SVC

In [36]:
model=SklearnClassifier(SVC(kernel='linear'))

In [37]:
#train model on testing data
model.train(training)

<SklearnClassifier(SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
    kernel='linear', max_iter=-1, probability=False, random_state=None,
    shrinking=True, tol=0.001, verbose=False))>

In [38]:
#test model on testing dataset
accuracy=nltk.classify.accuracy(model,testing)
print('SVC accuracy:',accuracy)

SVC accuracy: 0.788


#### Our model accuracy using SVM algorithm on training dataset using SVM classifier is 78.8%